### PR TITLE
strict numbers

### PR DESCRIPTION
**Pull request by orionz**
This PR is to match https://github.com/automerge/automerge/pull/371.  The json for scalar values now contains datatype 'int', 'uint', 'float32', 'float64'.

### CODE REVIEW

File: automerge-protocol/src/lib.rs
Comment (jeffa5) [inline automerge-protocol/src/lib.rs:316]: Should be f32

File: automerge-protocol/src/lib.rs
Comment (jeffa5) [inline automerge-protocol/src/lib.rs:324]: Should be f64

File: automerge-protocol/src/lib.rs
Comment (jeffa5) [inline automerge-protocol/src/lib.rs:348]: Shouldn't this be stricter? I would it expect to only return a `Some` for a `Uint`, similarly for the other `to_*` functions.

File: automerge-protocol/tests/serde_round_trip_proptest.rs
Comment (jeffa5) [inline automerge-protocol/tests/serde_round_trip_proptest.rs:126]: Needs a cleanup

**Review by jeffa5 — CHANGES_REQUESTED** [review]

**Comment by vedantroy** [comment]
Fixed the issue with the missing "preds" field.

File: automerge-protocol/src/serde_impls/op.rs
Comment (vedantroy) [inline automerge-protocol/src/serde_impls/op.rs:79]: Will this result in the `datatype` not being specified?

**Comment by vedantroy** [comment]
Looks like we are using an old version of the Rust compiler that thinks the [this syntax](https://github.com/rust-lang/rust/issues/54883) is experimental.

File: automerge-protocol/src/serde_impls/op.rs
Comment (vedantroy) [reply automerge-protocol/src/serde_impls/op.rs:79]: Here's an example:

```json
{
  "ops": [
    {
      "action": "set",
      "obj": "_root",
      "elemId": "_head",
      "values": [
        "hello",
        3
      ],
      "pred": []
    }
  ],
  "actor": "",
  "seq": 0,
  "startOp": 0,
  "time": 0,
  "message": null,
  "deps": []
}
```

Comment (vedantroy) [reply automerge-protocol/src/serde_impls/op.rs:79]: I think the correct way to solve this is to add a serializer for `ScalarValue`s.

**Comment by vedantroy** [comment]
@jeffa5 Is this ok to merge? (once I fix CI?)

**Comment by jeffa5** [comment]
@vedantroy There still seems to be a few discussions unresolved. As for CI, I'm not sure about the MSRV for us, we should probably decide on one and document it in the README.

I also think we should probably wait until the corresponding PR is merged in the js repo, then we won't have to rely on orion's fork for the interop tests.

File: automerge-protocol/src/serde_impls/op.rs
Comment (alexjg) [reply automerge-protocol/src/serde_impls/op.rs:79]: How is this intended to work in general? Is the idea that all the `values` would be the same datatype?

File: automerge-protocol/src/lib.rs
Comment (vedantroy) [reply automerge-protocol/src/lib.rs:348]: It seems like all the other `to_` functions are pretty lax?

Example:

```rust
    /// If this value can be coerced to an i64, return the i64 value
    pub fn to_i64(&self) -> Option<i64> {
        match self {
            ScalarValue::Int(n) => Some(*n),
            ScalarValue::Uint(n) => Some(*n as i64),
            ScalarValue::F32(n) => Some(*n as i64),
            ScalarValue::F64(n) => Some(*n as i64),
            ScalarValue::Counter(n) => Some(*n),
            ScalarValue::Timestamp(n) => Some(*n),
            _ => None,
        }
    }
```

Comment (jeffa5) [reply automerge-protocol/src/lib.rs:348]: Ok, we can leave it for now then.

File: automerge-protocol/src/serde_impls/op.rs
Comment (vedantroy) [reply automerge-protocol/src/serde_impls/op.rs]: @alexjg I believe so

Comment (vedantroy) [reply automerge-protocol/src/serde_impls/op.rs]: @alexjg Is there any reason in particular why values in a `MultiSet` should have the same type? Otherwise I can remove some code I added here.

Comment (alexjg) [reply automerge-protocol/src/serde_impls/op.rs]: Well if they are different types then we'll need to change the serialization logic to add a type tag to each value right? Otherwise how do you indicate that (for example) the first element in a list of numbers is a counter but the second is just an integer?

Comment (alexjg) [reply automerge-protocol/src/serde_impls/op.rs]: Note that I am using the word `datatype` quite specifically to mean the `datatype` tag that is used for single element inserts

Comment (ept) [reply automerge-protocol/src/serde_impls/op.rs]: I think MultiSet should focus on the case where all elements have the same datatype, since including a separate datatype tag for every value would add bloat, and the whole point of MultiSet is to be more compact. If the data contains values of different datatypes, they should be split into runs of MultiSet where each run only contains items of the same type (or use a plain Set if the run is of length 1).

**Review by jeffa5 — APPROVED** [review]
I think this looks ok.

Up to you whether we wait for the js side to be merged first or not for the interop tests.

**Comment by vedantroy** [comment]
@jeffa5 Can you run CI on this again (not sure why it isn't running?)

**Comment by jeffa5** [comment]
@vedantroy I've had a quick look but seems to be something above my access level, perhaps @orionz or @alexjg can see?

From looking at the requests section I can see

>  Could not authorize build request for automerge/automerge-rs. 

So it may be billing related or due to the move from travis-ci.org to .com which happened recently

File: automerge-frontend/src/state_tree/resolved_path.rs
Comment (alexjg) [inline automerge-frontend/src/state_tree/resolved_path.rs:944]: Could use `Vec::with_capacity(ops.len())` here.

File: automerge-protocol/src/lib.rs
Comment (alexjg) [inline automerge-protocol/src/lib.rs:333]: Can we move this and the `From` and `fmt::Display` impls below to `utility_impls`

**Comment by alexjg** [comment]
I think this will need to be re-added to the Automerge travis because of the .com -> .org migration, @orionz I think you are the only person with permission to do that.

**Comment by vedantroy** [comment]
@jeffa5 Getting some clippy errors about SmolStr & Arbitrary? Are these expected? Do I need to fix them?

**Comment by vedantroy** [comment]
There are 2 wasm interop tests failing. However, they are also failing on the main branch. So not related to this PR.

**Comment by vedantroy** [comment]
In the wasm interop tests, when checking out the JS repository, I'm getting a weird error "fatal: reference is not a tree: 14e01eebdcc7677a2316df5326ca104823c23183".

**Comment by vedantroy** [comment]
Ok, so there's an error in the wasm interop tests about a `TextDecoder` not being defined, I suspect we need to update our version of node.

**Comment by ept** [comment]
TextDecoder should be available in Node >= 12, which is the oldest currently supported version. I suggest updating to Node 16, which is the [currently active long-term-support version](https://nodejs.org/en/about/releases/).

**Pull request merged**
